ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Connection.scala

```diff
@@ -141,53 +141,59 @@ private[h2] class H2Connection[F[_]](
-              case Some(ix) => chunk.splitAt(ix)
-            }
+  def writeLoop: F[Unit] = {
+    def go(chunk: Chunk[H2Frame]): F[Unit] = state.get.flatMap { s =>
+      val fullDataSize = chunk.foldLeft(0) {
+        case (init, H2Frame.Data(_, data, _, _)) => init + data.size.toInt
+        case (init, _) => init
+      }
+      // println(s"Next Write Block Window - data: $fullDataSize window:${s.writeWindow} $s")
 
-            val bv = nonData.foldLeft(ByteVector.empty) { case (acc, frame) =>
-              acc ++ H2Frame.toByteVector(frame)
-            }
-            socket.isOpen.ifM(
-              socket.write(Chunk.byteVector(bv)) >>
-                nonData.traverse_(frame => logger.debug(s"$host:$port Write - $frame")),
-              new SocketException("Socket closed when attempting to write").raiseError,
-            ) >>
-              s.writeBlock.get.rethrow >>
-              go(after)
-          }
+      if (fullDataSize <= s.writeWindow && s.writeWindow > 0) {
+        val bv = chunk.foldLeft(ByteVector.empty) { case (acc, frame) =>
+          acc ++ H2Frame.toByteVector(frame)
         }
-        val firstGoAway = chunk.collectFirst { case g: H2Frame.GoAway =>
-          mapRef.get.flatMap { m =>
-            m.values.toList.traverse_(connection => connection.receiveGoAway(g))
-          } >> state.update(s => s.copy(closed = true))
+        state.update(s => s.copy(writeWindow = s.writeWindow - fullDataSize)) >>
+          socket.isOpen.ifM(
+            socket.write(Chunk.byteVector(bv)) >>
+              chunk.traverse_(frame => logger.debug(s"$host:$port Write - $frame")),
+            new SocketException("Socket closed when attempting to write").raiseError,
+          )
+      } else {
+        val (nonData, after) = chunk.indexWhere(_.isInstanceOf[H2Frame.Data]) match {
+          case None => (chunk, Chunk.empty[H2Frame])
+          case Some(ix) => chunk.splitAt(ix)
+        }
+
+        val bv = nonData.foldLeft(ByteVector.empty) { case (acc, frame) =>
+          acc ++ H2Frame.toByteVector(frame)
         }
-        firstGoAway.getOrElse(F.unit) >> go(chunk)
+        socket.isOpen.ifM(
+          socket.write(Chunk.byteVector(bv)) >>
+            nonData.traverse_(frame => logger.debug(s"$host:$port Write - $frame")),
+          new SocketException("Socket closed when attempting to write").raiseError,
+        ) >>
+          s.writeBlock.get.rethrow >>
+          go(after)
+      }
+    }
+
+    def firstGoAway(ch: Chunk[H2Frame]): F[Unit] =
+      ch.collectFirst { case g: H2Frame.GoAway =>
+        mapRef.get.flatMap { m =>
+          m.values.toList.traverse_(connection => connection.receiveGoAway(g))
+        } >> state.update(s => s.copy(closed = true))
+      }.sequence_
+
+    def loop: F[Unit] =
+      outgoing.tryTakeN(None).flatMap { (chs: List[Chunk[H2Frame]]) =>
+        val chunk = Chunk.concat(chs)
+        firstGoAway(chunk) >> go(chunk) >> loop
```

Comment:
Sorry for perhaps a silly question, but in the current implementation, all elements available in the queue `outgoing` will be dequeued and emitted as a single chunk once.  Although `QueueSource#tryTakeN(None)` also dequeues and emits all available elements, you're doing it in a loop. Could you please elaborate on why we need a loop here?

Reply:
Because the queue is constantly filling up, so after each "dequeue" operation is completed, the queue is still receiving new elements. 

A Stream intrinsically encodes that "pulling" loop, whereby each pull from the stream repeats the dequeue operation. What the PR is doing is cutting off that middle-man stream.